### PR TITLE
fix(grc20): handle errors in userToken.TransferFrom

### PR DESCRIPTION
**Pull request by wwqiu**
# Description

The spend allowance check is missing in TransferFrom function of userToken
``` go
func (t *userToken) TransferFrom(from, to std.Address, amount uint64) error {
	spender := std.GetOrigCaller()
	t.admin.spendAllowance(from, spender, amount)
	return t.admin.transfer(from, to, amount)
}
```

# How has this been tested?

test case (TestUserAllowance) in user_token_test.gno

### CODE REVIEW

File: examples/gno.land/p/demo/grc/grc20/user_token_test.gno [hunk 1]
```diff
@@ -1,9 +1,67 @@
...
+		expB:=b0+tt.spend 
+		// Check if balance equals expected value 
+		assert(t,b==expB ,"balance exp: %d,got %d",expB ,b)
+	}
+
+	// Try to transfer one token from owner to dest ,should fail because no allowance left 
+	err=dummyUser.TransferFrom(owner ,dest ,1)
+	assert(t,err!=nil ,"no allowance")
+
+}
```
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc20/user_token_test.gno:67]: Thank you for your contribution.

Could you run `gofmt` (or `gofumpt`) on this file please?

Thanks!

**Review by thehowl — APPROVED** [review]

File: examples/gno.land/p/demo/grc/grc20/user_token_test.gno (hunk 1 above)
Comment (wwqiu) [reply examples/gno.land/p/demo/grc/grc20/user_token_test.gno:67]: Thank you for your feedback.   

I have run gofmt on this file and pushed the changes.

File: examples/gno.land/p/demo/grc/grc20/user_token_test.gno [hunk 2]
```diff
@@ -1,9 +1,67 @@
...
+		// Calculate expected balance ,should be initial balance plus transfer amount
+		expB := b0 + tt.spend
+		// Check if balance equals expected value
+		assert(t, b == expB, "balance exp: %d,got %d", expB, b)
+	}
+
+	// Try to transfer one token from owner to dest ,should fail because no allowance left
+	err = dummyUser.TransferFrom(owner, dest, 1)
+	assert(t, err != nil, "no allowance")
+
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/grc/grc20/user_token_test.gno:66]: Can you delete this newline after the assertion check?
We have a linter for it, but it's not picking up `.gno` files yet

File: examples/gno.land/p/demo/grc/grc20/user_token_test.gno [hunk 3]
```diff
@@ -1,9 +1,67 @@
...
+		// Check if allowance equals expected value
+		assert(t, a == tt.exp, "allowance exp: %d,got %d", tt.exp, a)
+
+		// Get dest current balance
+		b, _ := dummyUser.BalanceOf(dest)
+		// Calculate expected balance ,should be initial balance plus transfer amount
+		expB := b0 + tt.spend
+		// Check if balance equals expected value
+		assert(t, b == expB, "balance exp: %d,got %d", expB, b)
+	}
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/grc/grc20/user_token_test.gno:61]: What do you think about extracting this test table into a separate unit test?
`TestUserAllowance` seems to test different things, and you can split it up so it's more clear and easier to manage in the future

**Review by zivkovicmilos — APPROVED** [review]
Thank you for the contribution 🙏 

I've left a few small comments we should discuss, otherwise it looks great 💯

File: examples/gno.land/p/demo/grc/grc20/user_token_test.gno (hunk 2 above)
Comment (wwqiu) [reply examples/gno.land/p/demo/grc/grc20/user_token_test.gno:66]: Sure

File: examples/gno.land/p/demo/grc/grc20/user_token_test.gno (hunk 3 above)
Comment (zivkovicmilos) [reply examples/gno.land/p/demo/grc/grc20/user_token_test.gno:61]: Sure, no problem, the test cases are dependent on approval - you can safely go ahead and rename the test to what you think is more appropriate. `TestApprove` sounds fine

Comment (wwqiu) [reply examples/gno.land/p/demo/grc/grc20/user_token_test.gno:61]: I have made and push the changes, please check again.

**Review by moul — APPROVED** [review]
LGTM

**Pull request merged**